File: src/19_remove_nth_node_from_end/remove_nth_from_end.py
"""
19. Remove Nth Node From End of List

Given the head of a linked list, remove the nth node from the end of the list and return its head.

Difficulty: Medium
"""

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        """
        Remove nth node from end using two-pointer technique with dummy node.

        Approach:
        1. Use dummy node to handle edge case of removing head
        2. Use two pointers with n gap between them
        3. Move both until fast reaches end
        4. Remove the node after slow pointer

        Time Complexity: O(L) where L is the length of the list
        Space Complexity: O(1) - only using pointers
        """
        
        dummy = ListNode()
        dummy.next = head
        slow = dummy
        fast = dummy
        
        while n > 0:
            fast = fast.next
            n -= 1
            
        while fast.next:
            slow = slow.next
            fast = fast.next
            
        slow.next = slow.next.next
        
        return dummy.next


# Helper functions for testing
def list_to_array(head: Optional[ListNode]) -> list:
    """Convert linked list to array for easy comparison."""
    result = []
    curr = head
    while curr:
        result.append(curr.val)
        curr = curr.next
    return result


def array_to_list(arr: list) -> Optional[ListNode]:
    """Convert array to linked list for testing."""
    if not arr:
        return None
    head = ListNode(arr[0])
    curr = head
    for val in arr[1:]:
        curr.next = ListNode(val)
        curr = curr.next
    return head


def test_remove_nth_from_end():
    solution = Solution()

    # Test case 1: Remove 2nd from end
    head1 = array_to_list([1, 2, 3, 4, 5])
    result1 = solution.removeNthFromEnd(head1, 2)
    print(f"Test 1: {list_to_array(result1)} (expected: [1, 2, 3, 5])")

    # Test case 2: Remove only node
    head2 = array_to_list([1])
    result2 = solution.removeNthFromEnd(head2, 1)
    print(f"Test 2: {list_to_array(result2)} (expected: [])")

    # Test case 3: Remove first of two nodes
    head3 = array_to_list([1, 2])
    result3 = solution.removeNthFromEnd(head3, 1)
    print(f"Test 3: {list_to_array(result3)} (expected: [1])")

    # Test case 4: Remove last node (head)
    head4 = array_to_list([1, 2])
    result4 = solution.removeNthFromEnd(head4, 2)
    print(f"Test 4: {list_to_array(result4)} (expected: [2])")

    # Test case 5: Remove from middle
    head5 = array_to_list([1, 2, 3, 4, 5])
    result5 = solution.removeNthFromEnd(head5, 3)
    print(f"Test 5: {list_to_array(result5)} (expected: [1, 2, 4, 5])")

    # Test case 6: Remove first node (last from end)
    head6 = array_to_list([1, 2, 3, 4, 5])
    result6 = solution.removeNthFromEnd(head6, 5)
    print(f"Test 6: {list_to_array(result6)} (expected: [2, 3, 4, 5])")

    # Test case 7: Remove last node
    head7 = array_to_list([1, 2, 3, 4, 5])
    result7 = solution.removeNthFromEnd(head7, 1)
    print(f"Test 7: {list_to_array(result7)} (expected: [1, 2, 3, 4])")


if __name__ == "__main__":
    test_remove_nth_from_end()

File: src/19_remove_nth_node_from_end/RemoveNthFromEnd.java
/**
 * 19. Remove Nth Node From End of List
 * 
 * Given the head of a linked list, remove the nth node from the end of the list
 * and return its head.
 * 
 * Difficulty: Medium
 */

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class RemoveNthFromEnd {
    /**
     * Remove nth node from end using two-pointer technique with dummy node.
     * 
     * Approach:
     * 1. Use dummy node to handle edge case of removing head
     * 2. Use two pointers with n gap between them
     * 3. Move both until fast reaches end
     * 4. Remove the node after slow pointer
     * 
     * Time Complexity: O(L) where L is the length of the list
     * Space Complexity: O(1) - only using pointers
     */
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode slow = dummy;
        ListNode fast = dummy;

        while (n > 0) {
            fast = fast.next;
            n--;
        }

        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }

        slow.next = slow.next.next;

        return dummy.next;
    }

    // Helper methods for testing
    private static int[] listToArray(ListNode head) {
        // Count nodes
        int count = 0;
        ListNode curr = head;
        while (curr != null) {
            count++;
            curr = curr.next;
        }

        // Convert to array
        int[] result = new int[count];
        curr = head;
        int i = 0;
        while (curr != null) {
            result[i++] = curr.val;
            curr = curr.next;
        }
        return result;
    }

    private static ListNode arrayToList(int[] arr) {
        if (arr == null || arr.length == 0) {
            return null;
        }
        ListNode head = new ListNode(arr[0]);
        ListNode curr = head;
        for (int i = 1; i < arr.length; i++) {
            curr.next = new ListNode(arr[i]);
            curr = curr.next;
        }
        return head;
    }

    private static String arrayToString(int[] arr) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < arr.length; i++) {
            sb.append(arr[i]);
            if (i < arr.length - 1) {
                sb.append(", ");
            }
        }
        sb.append("]");
        return sb.toString();
    }

    public static void main(String[] args) {
        RemoveNthFromEnd solution = new RemoveNthFromEnd();

        // Test case 1: Remove 2nd from end
        ListNode head1 = arrayToList(new int[] { 1, 2, 3, 4, 5 });
        ListNode result1 = solution.removeNthFromEnd(head1, 2);
        System.out.println("Test 1: " + arrayToString(listToArray(result1)) + " (expected: [1, 2, 3, 5])");

        // Test case 2: Remove only node
        ListNode head2 = arrayToList(new int[] { 1 });
        ListNode result2 = solution.removeNthFromEnd(head2, 1);
        System.out.println("Test 2: " + arrayToString(listToArray(result2)) + " (expected: [])");

        // Test case 3: Remove first of two nodes
        ListNode head3 = arrayToList(new int[] { 1, 2 });
        ListNode result3 = solution.removeNthFromEnd(head3, 1);
        System.out.println("Test 3: " + arrayToString(listToArray(result3)) + " (expected: [1])");

        // Test case 4: Remove last node (head)
        ListNode head4 = arrayToList(new int[] { 1, 2 });
        ListNode result4 = solution.removeNthFromEnd(head4, 2);
        System.out.println("Test 4: " + arrayToString(listToArray(result4)) + " (expected: [2])");

        // Test case 5: Remove from middle
        ListNode head5 = arrayToList(new int[] { 1, 2, 3, 4, 5 });
        ListNode result5 = solution.removeNthFromEnd(head5, 3);
        System.out.println("Test 5: " + arrayToString(listToArray(result5)) + " (expected: [1, 2, 4, 5])");

        // Test case 6: Remove first node (last from end)
        ListNode head6 = arrayToList(new int[] { 1, 2, 3, 4, 5 });
        ListNode result6 = solution.removeNthFromEnd(head6, 5);
        System.out.println("Test 6: " + arrayToString(listToArray(result6)) + " (expected: [2, 3, 4, 5])");

        // Test case 7: Remove last node
        ListNode head7 = arrayToList(new int[] { 1, 2, 3, 4, 5 });
        ListNode result7 = solution.removeNthFromEnd(head7, 1);
        System.out.println("Test 7: " + arrayToString(listToArray(result7)) + " (expected: [1, 2, 3, 4])");
    }
}
